sdks/java/core/src/main/java/org/apache/beam/sdk/util/MutationDetectors.java

```diff
@@ -148,13 +148,17 @@ private void verifyUnmodifiedThrowingCheckedExceptions() throws CoderException {
       // encodedObject, because the Coder may treat it differently.
       //
       // For example, an unbounded Iterable will be encoded in an unbounded way, but decoded into an
-      // ArrayList, which will then be re-encoded in a bounded format. So we really do need to
-      // encode-decode-encode retainedObject.
-      if (Arrays.equals(
-          CoderUtils.encodeToByteArray(coder, clonedOriginalObject),
-          CoderUtils.encodeToByteArray(coder, clonedPossiblyModifiedObject))) {
-        return;
-      }
+      // ArrayList, which will then be re-encoded in a bounded format. So we get a structural value
+      // from a coder and used that to check if the byte array[] are the same. The structuralValue()
+      // method of the Coder returns a StructuralByteArray object.
```

Comment:
This is not always the case; the Coder can understand that it has an object with a good `equals` method, and return the object directly.

I'm not sure if I agree that this is a desired change. We'll already have compared the final state of the original object to its clone in the original state, and the clone of the original object to the clone of original object in its final state, so we've already done what is appreciably a fast path, and now are checking the most generous possible "equality" in terms of java.